cephadm: validate fsid during cephadm shell command

Daniel-Pivonka opened the pull request:
--fsid would take any string and still drop you into a ceph shell that fully worked 

verifying the fsid matches the one in the ceph conf

Fixes: https://tracker.ceph.com/issues/49724
Signed-off-by: Daniel Pivonka <dpivonka@redhat.com>

sebastian-philipp commented on src/cephadm/cephadm:
```diff
@@ -4237,6 +4237,15 @@ def command_run(ctx):
 @infer_image
 def command_shell(ctx):
     # type: (CephadmContext) -> int
+    (config, _) = get_config_and_keyring(ctx)
+    fsid_match = False
+    for c in config.split('\n'):
+        if 'fsid = '+ctx.fsid == c.strip():
+            fsid_match = True
+
+    if not fsid_match:
+        raise Error('fsid does not match ceph conf')
```
This is the **exact** piece of code that we need in order to close https://tracker.ceph.com/issues/48368

Mind if you make this into a function and call it from check-host as well?

Daniel-Pivonka replied:
i moved the code into a function but im not able to figure out a way to make it work in check-host. 

check-host is called when adding all hosts so there wont be a /etc/ceph/ceph.conf on a new hostbeing added to the cluster to check 

additionally fsid is not passed to the check_host call so theres no guaranteed fsid to check for.

there is no guarantee of a conf being in /etc/ceph on every host in the cluster unless the 'manage_etc_ceph_ceph_conf' is set.

i could check for that flag and then try to match the fsid in /etc/ceph/ceph.conf to one of the fsids in /var/lib/ceph(thats where infer_fsid is looking for a list of fsids) but im not seeing the value of having that check?

sebastian-philipp commented:
```
2021-03-16T09:49:01.988 INFO:tasks.workunit.client.0.smithi053.stdout:ceph version 16.1.0-736-g8191ac78 (8191ac7873f4253a74b5e91d2a1344df8e2a591c) pacific (rc)
2021-03-16T09:49:02.013 INFO:tasks.workunit.client.0.smithi053.stderr:+ sudo CEPHADM_IMAGE=quay.ceph.io/ceph-ci/ceph:pacific /home/ubuntu/cephtest/clone.client.0/qa/workunits/cephadm/../../../src/cephadm/cephadm version
2021-03-16T09:49:02.013 INFO:tasks.workunit.client.0.smithi053.stderr:+ grep 'ceph version 16'
2021-03-16T09:49:03.639 INFO:tasks.workunit.client.0.smithi053.stdout:ceph version 16.1.0-736-g8191ac78 (8191ac7873f4253a74b5e91d2a1344df8e2a591c) pacific (rc)
2021-03-16T09:49:03.640 INFO:tasks.workunit.client.0.smithi053.stderr:+ grep 'ceph version'
2021-03-16T09:49:03.640 INFO:tasks.workunit.client.0.smithi053.stderr:+ sudo /home/ubuntu/cephtest/clone.client.0/qa/workunits/cephadm/../../../src/cephadm/cephadm --image quay.ceph.io/ceph-ci/ceph:master version
2021-03-16T09:49:59.455 INFO:tasks.workunit.client.0.smithi053.stdout:ceph version 17.0.0-2061-gc62bd5e0 (c62bd5e05ea72e884b14fd3f2555fbc9d4755b33) quincy (dev)
2021-03-16T09:49:59.456 INFO:tasks.workunit.client.0.smithi053.stderr:+ systemctl status docker
2021-03-16T09:49:59.466 INFO:tasks.workunit.client.0.smithi053.stderr:+ sudo /home/ubuntu/cephtest/clone.client.0/qa/workunits/cephadm/../../../src/cephadm/cephadm --image quay.ceph.io/ceph-ci/ceph:master --docker version
2021-03-16T09:49:59.466 INFO:tasks.workunit.client.0.smithi053.stderr:+ grep 'ceph version'
2021-03-16T09:50:27.897 INFO:tasks.workunit.client.0.smithi053.stdout:ceph version 17.0.0-2061-gc62bd5e0 (c62bd5e05ea72e884b14fd3f2555fbc9d4755b33) quincy (dev)
2021-03-16T09:50:27.898 INFO:tasks.workunit.client.0.smithi053.stderr:+ sudo /home/ubuntu/cephtest/clone.client.0/qa/workunits/cephadm/../../../src/cephadm/cephadm --image quay.ceph.io/ceph-ci/ceph:master shell --fsid 00000000-0000-0000-0000-0000deadbeef -- ceph -v
2021-03-16T09:50:27.899 INFO:tasks.workunit.client.0.smithi053.stderr:+ grep 'ceph version'
2021-03-16T09:50:28.226 INFO:tasks.workunit.client.0.smithi053.stderr:ERROR: fsid does not match ceph conf
```

https://pulpito.ceph.com/swagner-2021-03-16_08:50:44-rados:cephadm-wip-swagner-testing-2021-03-12-1502-distro-basic-smithi/5970558

sebastian-philipp commented:
https://pulpito.ceph.com/swagner-2021-03-16_08:50:44-rados:cephadm-wip-swagner-testing-2021-03-12-1502-distro-basic-smithi/

failures: 

* https://tracker.ceph.com/issues/49633
* e2e test 
* #40015
* test_cephfs_mirror

liewegas commented on src/cephadm/cephadm:
```diff
@@ -4232,11 +4232,25 @@ def command_run(ctx):
 ##################################
 
 
+def fsid_conf_match(ctx):
```
fsid_conf_mismatch

liewegas commented on src/cephadm/cephadm:
```diff
@@ -4232,11 +4232,25 @@ def command_run(ctx):
+    if config:
+        for c in config.split('\n'):
+            if 'fsid = ' + ctx.fsid == c.strip():
+                fsid_match = True
```
if ... != ... then return False

liewegas commented on src/cephadm/cephadm:
```diff
@@ -4232,11 +4232,25 @@ def command_run(ctx):
+        for c in config.split('\n'):
+            if 'fsid = ' + ctx.fsid == c.strip():
+                fsid_match = True
+    return fsid_match
```
return True

Daniel-Pivonka commented:
changed the logic around per sages suggestions. also added a few checks to make sure files exist before trying to open them to avoid trackbacks that are possible in rare situations. test should pass now

Daniel-Pivonka commented:
jenkins test make check

ktdreyer commented on src/cephadm/cephadm:
```diff
@@ -1656,6 +1656,8 @@ def infer_config(func):
                                                              name)
         if config:
+            if not os.path.isfile(config):
+                raise Error('Cannot infer a config')
```
This error message is hard to understand. Can we raise `FileNotFoundError` instead and show the full path instead?

ktdreyer replied:
Why don't we catch `FileNotFoundError`?

Daniel-Pivonka replied:
your right this check is not really needed. the try execpt in get_config_and_keyring will catch this.

ktdreyer commented on src/cephadm/cephadm:
```diff
@@ -2217,12 +2219,16 @@ def get_config_and_keyring(ctx):
 
     if 'config' in ctx and ctx.config:
+        if not os.path.isfile(ctx.config):
+            raise Error('config file %s not found' % ctx.config)
```
the next line calls `open()`, and that will already raise `FileNotFoundError`. No need to Look Before We Leap here

Daniel-Pivonka replied:
ive put a try except around the open() call and raise an "Error"
![image](https://user-images.githubusercontent.com/22037319/112187855-d785d080-8bd8-11eb-86d4-4acc6b35be74.png)
same reasoning as above, i think its better to not show the full traceback, the message is enough for the user

ktdreyer commented on src/cephadm/cephadm:
```diff
@@ -2217,12 +2219,16 @@ def get_config_and_keyring(ctx):
         keyring = '[%s]\n\tkey = %s\n' % (ctx.name, ctx.key)
     elif 'keyring' in ctx and ctx.keyring:
+        if not os.path.isfile(ctx.keyring):
+            raise Error('keyring file %s not found' % ctx.keyring)
```
Again, the next line calls `open()` which will raise `FileNotFoundError`.

Daniel-Pivonka replied:
same as above

ktdreyer commented on src/cephadm/cephadm:
```diff
@@ -4232,11 +4238,25 @@ def command_run(ctx):
 def command_shell(ctx):
     # type: (CephadmContext) -> int
+    if not fsid_conf_mismatch(ctx):
+        raise Error('fsid does not match ceph conf')
```
The boolean handling is confusing to me. "`if not ... mismatch`, then raise an error?" Can we rename this method `fsid_conf_match()`, since we raise the error when the return value is `False`?

liewegas replied:
This is deliberate, since we don't require a match (the fsid can be omitted)... we just want to bail out if it is present AND it is a mismatch.

liewegas replied:
Oh, the function should flip true/false around, though, you're right!

liewegas commented on src/cephadm/cephadm:
```diff
@@ -4232,11 +4238,25 @@ def command_run(ctx):
+            if 'fsid = ' in c.strip():
+                if 'fsid = ' + ctx.fsid != c.strip():
+                    return False
+    return True
```
the return values should be swapped... True if fsid is present and a mismatch

jmolmo commented on src/cephadm/cephadm:
```diff
@@ -4232,11 +4238,25 @@ def command_run(ctx):
 def command_shell(ctx):
     # type: (CephadmContext) -> int
+    if fsid_conf_mismatch(ctx):
+        raise Error('fsid does not match ceph conf')
```
The message is completely clear for us .. but maybe a little bit cryptic for an user..

What about: "fsid provided does not match with fsid found in ceph conf"
